### PR TITLE
OWLAP-51: Effectivetime restore updates

### PR DESCRIPTION
**Pull request by zoliMolnar**
This pull request adds functionality to restore effective time on members and components.

### CODE REVIEW

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+						&& memberToRestore.getValueId().equals((String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_VALUE_ID));
+			} else if (componentToRestore instanceof SnomedConcreteDataTypeRefSetMember) {
+				final SnomedConcreteDataTypeRefSetMember memberToRestore = (SnomedConcreteDataTypeRefSetMember) componentToRestore;
+				return memberToRestore.isActive() == previousMember.isActive() 
+						&& memberToRestore.getModuleId().equals(previousMember.getModuleId())
+						&& memberToRestore.getSerializedValue().equals(previousMember.getProperties().get(SnomedRf2Headers.FIELD_VALUE))
+						&& memberToRestore.getCharacteristicTypeId().equals(previousMember.getProperties().get(SnomedRf2Headers.FIELD_CHARACTERISTIC_TYPE_ID))
+						&& memberToRestore.getTypeId().equals(previousMember.getProperties().get(SnomedRf2Headers.FIELD_TYPE_ID))
+						&& memberToRestore.getGroup() == ((Integer )previousMember.getProperties().get(SnomedRf2Headers.FIELD_RELATIONSHIP_GROUP)).intValue();
+			} else if(componentToRestore instanceof SnomedDescriptionTypeRefSetMember) {
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java:173]: Fix spacing, please! :)

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+				return memberToRestore.isActive() == previousMember.isActive() && memberToRestore.getModuleId().equals(previousMember.getModuleId());
+			} else if (componentToRestore instanceof SnomedMRCMAttributeDomainRefSetMember) {
+				final SnomedMRCMAttributeDomainRefSetMember memberToRestore = (SnomedMRCMAttributeDomainRefSetMember) componentToRestore;
+				final Boolean previousdGrouped = ClassUtils.checkAndCast(previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_GROUPED), Boolean.class);		
+				final String previousAttributeCardinality = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_ATTRIBUTE_CARDINALITY);
+				final String previousAtributeInGroupCardinality = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_ATTRIBUTE_IN_GROUP_CARDINALITY);
+				final String previousRuleStrengthId = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_RULE_STRENGTH_ID);
+				final String previousContentTypeId = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_CONTENT_TYPE_ID);
+				
+				if (previousdGrouped != null && previousdGrouped.booleanValue() != memberToRestore.isGrouped()) {
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java]: Typo!

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+					return false;
+				}
+
+				if (previousAttributeCardinality != null && !previousAttributeCardinality.equals(memberToRestore.getAttributeCardinality())) {
+					return false;
+				}
+
+				if (previousAtributeInGroupCardinality != null && !previousAtributeInGroupCardinality.equals(memberToRestore.getAttributeInGroupCardinality())) {
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java]: Typo!

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+					componentsByType.put(object.eClass().getInstanceClass(), object);
+				}
+				if (object instanceof SnomedRefSetMember && ((SnomedRefSetMember) object).isReleased() && !((SnomedRefSetMember) object).isSetEffectiveTime()) {
+					componentsByType.put(SnomedRefSetMember.class, object);
+				}
+			});
+		if (componentsByType.isEmpty())	return;
+		
+		final List<String> branchesForPreviousVersion = getAvailableVersionPaths();
+		for (String branch : branchesForPreviousVersion) {
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java:101]: This list of branches provided here has the following restrictions:
- the previous version of the component **must** exist on at least one branch. if we can't find it on any of the branches then we should propagate that upwards. Thoughts @cmark?
- it is **not** guaranteed that the component exists on all branches
- if we find the component on a branch then we **must not** look further on the next available branch

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+				final String previousParentDomain = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_PARENT_DOMAIN);
+				final String previousProximalPrimitiveConstraint = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_PROXIMAL_PRIMITIVE_CONSTRAINT);
+				final String previousProximalPrimitiveRefinement = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_PROXIMAL_PRIMITIVE_REFINEMENT);
+				final String previousDomainTemplateForPrecoordination = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_DOMAIN_TEMPLATE_FOR_PRECOORDINATION);
+				final String previousDomainTemplateForPostcoordination = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_DOMAIN_TEMPLATE_FOR_POSTCOORDINATION);
+				final String previousEditorialGuideReference = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_EDITORIAL_GUIDE_REFERENCE);
+
+				if (previousDomainConstraint != null && !previousDomainConstraint.equals(memberToRestore.getDomainConstraint())) {
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java]: Use `java.util.Objects.equals()` to eliminate unnecessary `null` checks.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+
+					if (!Objects.equal(previousTargetEffectiveDate, memberToRestore.getTargetEffectiveTime())) {
+						return false;
+					}
+				}
+
+				return memberToRestore.isActive() == previousMember.isActive() && memberToRestore.getModuleId().equals(previousMember.getModuleId());
+			} else if (componentToRestore instanceof SnomedMRCMAttributeDomainRefSetMember) {
+				final SnomedMRCMAttributeDomainRefSetMember memberToRestore = (SnomedMRCMAttributeDomainRefSetMember) componentToRestore;
+				final Boolean previousdGrouped = ClassUtils.checkAndCast(previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_GROUPED), Boolean.class);
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java]: Typo!

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+						return false;
+					}
+				}
+
+				return memberToRestore.isActive() == previousMember.isActive() && memberToRestore.getModuleId().equals(previousMember.getModuleId());
+			} else if (componentToRestore instanceof SnomedMRCMAttributeDomainRefSetMember) {
+				final SnomedMRCMAttributeDomainRefSetMember memberToRestore = (SnomedMRCMAttributeDomainRefSetMember) componentToRestore;
+				final Boolean previousdGrouped = ClassUtils.checkAndCast(previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_GROUPED), Boolean.class);		
+				final String previousAttributeCardinality = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_ATTRIBUTE_CARDINALITY);
+				final String previousAtributeInGroupCardinality = (String) previousMember.getProperties().get(SnomedRf2Headers.FIELD_MRCM_ATTRIBUTE_IN_GROUP_CARDINALITY);
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java]: Typo!

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+		if (object instanceof Component) {
+			final Component component = (Component) object;
+			return component.getId();
+		} else if (object instanceof SnomedRefSetMember) {
+			final SnomedRefSetMember member = (SnomedRefSetMember) object;
+			return member.getUuid();
+		}
+		
+		// XXX: Should never happen
+		throw new IllegalStateException();
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java]: In this case, throw an `UnsupportedOperationException` or `IllegalArgumentException` with a proper message.

**Review by cmark — CHANGES_REQUESTED** [review]

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+			if (componentToRestore.eClass().equals(SnomedRefSetPackage.Literals.SNOMED_REF_SET_MEMBER)) {
+				// Simple type reference set
+				final SnomedRefSetMember memberToRestore = (SnomedRefSetMember) componentToRestore;
+				return memberToRestore.isActive() == previousMember.isActive() && memberToRestore.getModuleId().equals(previousMember.getModuleId());
+			} else {
+				return false;
+			}
+		} else {
+			throw new UnexpectedTypeException("Unexpected member type '" + componentToRestore.getClass() + "'.");
+		}
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java:374]: That's quite a long method but we are on a time schedule now, so I guess it should be ok.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedEditingContext.java [hunk 10]
```diff
@@ -791,8 +791,11 @@ public void preCommit() {
 				}
 			}
 		}
+		
+		new EffectiveTimeRestorer(getBranch()).restoreEffectiveTimes(transaction.getChangeSetData().getChangedObjects().stream()
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedEditingContext.java]: Why not instantiate the restorer in the constructor and just pass the branch to the method?

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java (hunk 4 above)
Comment (nagyo) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java:101]: I might even change the two for loops here. First iterate over types and then branches?

**Review by nagyo — CHANGES_REQUESTED** [review]

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java (hunk 1 above)
Comment (zoliMolnar) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/EffectiveTimeRestorer.java:173]: This doesn't look the same inside my IDE.

**Review by cmark — APPROVED** [review]
LGTM!

**Pull request merged**